strapi/strapi
# Upgrade to lerna 5 to speedup build

### What does it do?

Update `lerna` version `4` to `5` to leverage caching feature in `lerna` 5 (powered by `nx`). See **More context** for more

### Why is it needed?

- Cache build folders when run `yarn build`
- Increase DX

## Performance gain

The following table shows the performance gain when upgrading `lerna` to version `5`. We can see that the build time cut by half in the second build. It takes under 0.5 seconds for the third time.


| Command              | Lerna 4 | Lerna 5 |
|----------------------|---------|---------|
| yarn build (1st run) | 12.5s    | 12.5s   |
| yarn build (2nd run) | 10.6s    | 6.4s    |
| yarn build (3rd run) | 10.6s    | 0.45s   |

*(Number recorded when running on a Macbook Air M1 2020, I believe the number in Intel-based computers will be more impressive)*

# Some screenshots

`yarn build`, powered by Nx, hit the cache
<img width="677" alt="Lerna (powered by Nx)   Running target build for 2 projects" src="https://user-images.githubusercontent.com/8603085/176673081-f0f61dde-11a2-45cb-852c-f8d2a04be948.png">

Running `yarn build` the third time takes under 0.5 seconds, thanks to caching feature.

<img width="747" alt="Done in 0.48s" src="https://user-images.githubusercontent.com/8603085/176673271-c583f8e2-5106-48b0-ab8c-d7f96c68eab1.png">


### How to test it?

- Make sure `lerna@5` is installed `yarn`.
- Build: `yarn build`
- Rebuild the first time: `yarn build`. The build time the is cut by half
- Rebuild the second time: `yarn build`. The build time is under 0.5 seconds ⚡️

### More context
- [Nx](https://nx.dev) recently took over the stewardship of [Lerna.js](https://lerna.js.org/), with a lot of improvements regarding performance.
- [Lerna is dead — Long Live Lerna](https://blog.nrwl.io/lerna-is-dead-long-live-lerna-61259f97dbd9)
- [Lerna 5.1 — New website, new guides, new Lerna example repo, distributed caching support and speed!](https://blog.nrwl.io/lerna-5-1-new-website-new-guides-new-lerna-example-repo-distributed-caching-support-and-speed-64d66410bec7)
- [How Lerna just got 10x faster!](https://blog.nrwl.io/lerna-used-to-walk-now-it-can-fly-eab7a0fe7700)
- [Integrating Nx and Lerna](https://nx.dev/migration/lerna-and-nx)

cc: @gu-stav
